File: Youtube_shots_Maker/lib/api/publisher.ts
import { api } from "./client";

// ─── 타입 ───────────────────────────────────────────────────

export interface PublishTarget {
  account_id: string;
  scheduled_at?: string; // ISO 8601
}

export interface SchedulePublishRequest {
  script_id: string;
  publish_targets: PublishTarget[];
  commerce_link?: string;
  proxy_id?: string;
}

export interface FaqItem {
  question: string;
  answer: string;
}

export interface GeoMetadata {
  title: string;
  description: string;
  tags: string[];
  context_summary: string;
  faq_schema: FaqItem[];
}

export type JobStatus =
  | "pending"
  | "scheduled"
  | "uploading"
  | "published"
  | "failed";

export interface PublishJob {
  id: string;
  video_id: string | null;
  script_id: string | null;
  publish_account_id: string | null;
  scheduled_at: string | null;
  published_at: string | null;
  status: JobStatus;
  platform_video_id: string | null;
  commerce_link: string | null;
  geo_metadata: GeoMetadata | null;
  error_msg: string | null;
  created_at: string;
  account_alias: string | null;
  account_platform: string | null;
  account_display_name: string | null;
}

// ─── API 함수 ────────────────────────────────────────────────

export const publisherApi = {
  schedule: (body: SchedulePublishRequest) =>
    api.post<{ job_ids: string[] }>("/api/publisher/schedule", body),

  listJobs: (account_id?: string) => {
    const qs = account_id ? `?account_id=${account_id}` : "";
    return api.get<PublishJob[]>(`/api/publisher/jobs${qs}`);
  },

  geoMetadata: (script_id: string) =>
    api.post<GeoMetadata>("/api/publisher/geo-metadata", { script_id }),
};

File: Youtube_shots_Maker/backend/models/publisher.py
"""
models/publisher.py
Publisher Matrix — Pydantic 스키마 (Phase 3)
"""

from __future__ import annotations

from datetime import datetime
from typing import Optional
from pydantic import BaseModel


# ─────────────────────────────────────────────────────────────
# 요청 스키마
# ─────────────────────────────────────────────────────────────

class PublishTarget(BaseModel):
    account_id: str
    scheduled_at: Optional[datetime] = None   # None → 즉시 배포


class SchedulePublishRequest(BaseModel):
    script_id: str
    publish_targets: list[PublishTarget]
    commerce_link: Optional[str] = None
    proxy_id: Optional[str] = None


class GeoMetadataRequest(BaseModel):
    script_id: str


# ─────────────────────────────────────────────────────────────
# GEO 메타데이터
# ─────────────────────────────────────────────────────────────

class FaqItem(BaseModel):
    question: str
    answer: str


class GeoMetadataResponse(BaseModel):
    title: str
    description: str
    tags: list[str]
    context_summary: str
    faq_schema: list[FaqItem]


# ─────────────────────────────────────────────────────────────
# 응답 스키마
# ─────────────────────────────────────────────────────────────

class SchedulePublishResponse(BaseModel):
    job_ids: list[str]


class PublishJobResponse(BaseModel):
    id: str
    video_id: Optional[str] = None
    script_id: Optional[str] = None
    publish_account_id: Optional[str] = None
    scheduled_at: Optional[datetime] = None
    published_at: Optional[datetime] = None
    status: str
    platform_video_id: Optional[str] = None
    commerce_link: Optional[str] = None
    geo_metadata: Optional[dict] = None
    error_msg: Optional[str] = None
    created_at: datetime
    # 계정 정보 (JOIN)
    account_alias: Optional[str] = None
    account_platform: Optional[str] = None
    account_display_name: Optional[str] = None
